subprojects/core/src/main/java/org/gradle/deployment/internal/DefaultDeploymentRegistry.java

```diff
@@ -84,11 +151,28 @@ public void stop() {
             handles.clear();
             lock.unlock();
         }
+        pendingChangesManager.removeListener(this);
     }
 
     private void failIfStopped() {
         if (stopped) {
             throw new IllegalStateException("Cannot modify deployment handles once the registry has been stopped.");
         }
     }
+
+    private static class PendingChanges {
+        private int pendingChanges;
+
+        void changesMade() {
+            pendingChanges++;
+        }
+
+        void changesIncorporated() {
+            pendingChanges = Math.max(0, pendingChanges-1);
```

Comment:
I don't like the way this assumes that we will get only 1 `changesMade` signal per build, and this isn't part of the contract (it's down to the fact that we have a 'SingleFirePendingChangesListener`)